node/src/proxy_server/mod.rs

```diff
@@ -4157,6 +4732,367 @@ mod tests {
+                return_route_id: 1234,
+                expected_services: expected_services.clone(),
+                protocol: ProxyProtocol::HTTP,
+                hostname_opt: Some("server.com".to_string()),
+            },
+        );
+        let message_resolver = RouteQueryResponseResolverMock::default()
+            .resolve_message_params(&resolve_message_params_arc);
+        let message_resolver_factory = RouteQueryResponseResolverFactoryMock::default()
+            .make_result(Box::new(message_resolver));
+        subject.inbound_client_data_helper_opt = Some(Box::new(IBCDHelperReal {
+            factory: Box::new(message_resolver_factory),
+        }));
+        let subject_addr: Addr<ProxyServer> = subject.start();
+        let dns_resolve_failure = DnsResolveFailure_0v1::new(stream_key);
+        let expired_cores_package: ExpiredCoresPackage<DnsResolveFailure_0v1> =
+            ExpiredCoresPackage::new(
+                SocketAddr::from_str("1.2.3.4:1234").unwrap(),
+                Some(make_wallet("irrelevant")),
+                return_route_with_id(cryptde, 1234),
+                dns_resolve_failure.into(),
+                0,
+            );
+        let mut peer_actors = peer_actors_builder()
+            .neighborhood(neighborhood_mock)
+            .build();
+        peer_actors.proxy_server = ProxyServer::make_subs_from(&subject_addr);
+
+        subject_addr.try_send(BindMessage { peer_actors }).unwrap();
+        subject_addr.try_send(expired_cores_package).unwrap();
+
+        subject_addr
+            .try_send(AssertionsMessage {
+                assertions: Box::new(move |proxy_server: &mut ProxyServer| {
+                    let retry = proxy_server.dns_failure_retries.get(&stream_key).unwrap();
+                    assert_eq!(retry.retries_left, 2);
+                }),
+            })
+            .unwrap();
+        let before = SystemTime::now();
+        system.run();
+        let after = SystemTime::now();
+        let mut resolve_message_params = resolve_message_params_arc.lock().unwrap();
+        let (transmit_to_hopper_args, route_query_message_response) =
+            resolve_message_params.remove(0);
+        let args = transmit_to_hopper_args;
+
+        assert!(resolve_message_params.is_empty());
+        assert_eq!(args.payload, client_payload);
+        assert_eq!(args.source_addr, socket_addr);
+        assert!(before <= args.timestamp && args.timestamp <= after);
+        assert!(args.retire_stream_key_sub_opt.is_none());
+        assert_eq!(args.is_decentralized, true);
+        assert_eq!(
+            route_query_message_response.unwrap().unwrap(),
+            route_query_response_expected
+        );
+    }
+
+    #[test]
+    #[should_panic(
+        expected = "No entry found inside dns_failure_retries hashmap for the stream_key: +dKB2Lsh3ET2TS/J/cexaanFQz4"
+    )]
+    fn handle_dns_resolve_failure_panics_when_there_is_no_entry_in_the_hashmap_for_the_stream_key()
```

Comment:
Is this a duplicate test?

Reply:
We're panicking at two different places in case we're not able to find an entry inside the Hashmap.

The two places are:

1) Handler for `RouteResultMessage` - This message contains information from Neighbourhood intended for ProxyServer.
2) Inside `handle_dns_resolve_failure()` - This is triggered when we receive an ExpiredCoresPackage containing a DNS Failure.

I do think that the second case is a security risk and it must be changed from a panic to an error log.